src/libstd/sys/sgx/abi/usercalls/mod.rs

```diff
@@ -149,7 +150,28 @@ pub fn exit(panic: bool) -> ! {
 
 /// Usercall `wait`. See the ABI documentation for more information.
 #[unstable(feature = "sgx_platform", issue = "56975")]
-pub fn wait(event_mask: u64, timeout: u64) -> IoResult<u64> {
+pub fn wait(event_mask: u64, mut timeout: u64) -> IoResult<u64> {
+    if timeout != WAIT_NO && timeout != WAIT_INDEFINITE {
+        // We don't want people to rely on accuracy of timeouts to make
+        // security decisions in an SGX enclave. That's why we add a random
+        // amount not exceeding +/- 10% to the timeout value to discourage
+        // people from relying on accuracy of timeouts while providing a way
+        // to make things work in other cases. Note that in the SGX threat
+        // model the enclave runner which is serving the wait usercall is not
+        // trusted to ensure accurate timeouts.
+        let base = cmp::max(1, timeout / 10) * 2 + 1;
+        let zero = base / 2;
+        match rdrand64() % base {
+            jitter if jitter > zero => {
+                timeout = timeout.checked_add(jitter - zero).unwrap_or(timeout)
+            }
+            jitter if jitter < zero => {
+                timeout = timeout.checked_sub(zero - jitter).unwrap_or(timeout)
+            }
+            _ => {}
+        };
+        timeout = cmp::min(u64::MAX - 1, cmp::max(1, timeout));
```

Comment:
This algorithm is a little hard to understand and I don't think it's necessary to precisely handle all the corner cases for impractically large timeouts. Something like this should work fine for all timeouts <= 2⁶²ns:
```
if let Ok(timeout_signed) = i64::try_from(timeout) {
    let tenth = timeout_signed / 10;
    let deviation = (rdrand64() as i64).checked_rem(tenth * 2).map_or(0, |rand| rand - tenth);
    timeout = timeout_signed.saturating_add(deviation) as _;
}
```

Reply:
That does not seem to work properly: https://play.rust-lang.org/?version=stable&mode=debug&edition=2018&gist=0f5bcdf8fe80f313cb202c6eab85b105

Reply:
Right, it's even simpler than I said before:
```
    if let Ok(timeout_signed) = i64::try_from(timeout) {
        let tenth = timeout_signed / 10;
        let deviation = (rdrand64() as i64).checked_rem(tenth).unwrap_or(0);
        timeout = timeout_signed.saturating_add(deviation) as _;
    }
```